File: BCKEND/JAVA/BancoFinanceiro/src/Model/Transacao.java
package Model;

import java.util.Date;

enum TipoTransacao {
	Depósito, Saque, Pagamento, Transferência
}

public class Transacao {

	private TipoTransacao tipo;
	private Date data;
	private double valor;
	private Cliente clienteTransferencia;
	private String infoPag;
	private char tipoValor;

	public Transacao(TipoTransacao tipo, double valor, char tipoValor) {
		super();
		this.tipo = tipo;
		this.data = new Date();
		this.valor = valor;
		this.clienteTransferencia = null;
		this.infoPag = null;
		this.tipoValor = tipoValor;
	}

	public Transacao(TipoTransacao tipo, double valor, Cliente clienteTransferencia, char tipoValor) {
		super();
		this.tipo = tipo;
		this.data = new Date();
		this.valor = valor;
		this.clienteTransferencia = clienteTransferencia;
		this.infoPag = null;
		this.tipoValor = tipoValor;
	}

	public Transacao(TipoTransacao tipo, double valor, String infoPag, char tipoValor) {
		super();
		this.tipo = tipo;
		this.data = new Date();
		this.valor = valor;
		this.clienteTransferencia = null;
		this.infoPag = infoPag;
		this.tipoValor = tipoValor;
	}

	public TipoTransacao getTipo() {
		return tipo;
	}

	public Date getData() {
		return data;
	}

	public double getValor() {
		return valor;
	}

	public Cliente getClienteTransferencia() {
		return clienteTransferencia;
	}

	public char getTipoValor() {
		return tipoValor;
	}

	public String getInfoPagamento() {
		return infoPag;
	}

	@Override
	public String toString() {
		if (clienteTransferencia != null) {
			return data.getDate() + "/" + (data.getMonth()+1) + "/" + 
					(data.getYear()+1900) + "\t" + tipo + "\t"
					+ clienteTransferencia.getNome() + "\t" + tipoValor + 
					" R$ " + valor;
		} else if(infoPag != null) {
			return data.getDate() + "/" + (data.getMonth()+1) + "/" + 
					(data.getYear()+1900) + "\t" + tipo + "\t"
					+ infoPag + "\t" + tipoValor + " R$ " + valor;
		} else {
			return data.getDate() + "/" + (data.getMonth()+1) + "/" + 
					(data.getYear()+1900) + "\t" + tipo + "\t"
					+ tipoValor + " R$ " + valor;
		}
	}

}

File: BCKEND/TypeScript/BancoFinanceiro/src/Model/Transacao.ts
import { Cliente } from "./Cliente";

export enum TipoTransacao {
    DEPOSITO = "Depósito",
    SAQUE = "Saque",
    TRANSFERENCIA = "Transferência",
    PAGAMENTO = "Pagamento"
}

export class Transacao {
    private tipo: TipoTransacao;
    private valor: number;
    private data: Date;
    private clienteTransferencia?: Cliente;
    private infoPagamento?: string;
    private tipoValor: string;

    constructor(tipo: TipoTransacao, valor: number, tipoValor: string, clienteTransferencia?: Cliente, infoPagamento?: string) {
        this.tipo = tipo;
        this.valor = valor;
        this.data = new Date();
        if(clienteTransferencia) this.clienteTransferencia = clienteTransferencia;
        if(infoPagamento) this.infoPagamento = infoPagamento;
        this.tipoValor = tipoValor;
    }

    public getTipo(): TipoTransacao {
        return this.tipo;
    }

    public getValor(): number {
        return this.valor;
    }

    public getData(): Date {
        return this.data;
    }

    public getClienteTransferencia(): Cliente | undefined {
        return this.clienteTransferencia;
    }

    public getTipoValor(): string {
        return this.tipoValor;
    }

    public getInfoPagamento(): string | undefined {
        return this.infoPagamento;
    }

    public toString(): string {
        let transacaoStr = `Data: ${this.data.toLocaleString()}, Tipo: ${this.tipo}, Valor: ${this.tipoValor} ${this.valor.toFixed(2)}`;
        if (this.clienteTransferencia) {
            transacaoStr += `, Cliente Transferência: ${this.clienteTransferencia.getNome()}`;
        }
        if (this.infoPagamento) {
            transacaoStr += `, Info Pagamento: ${this.infoPagamento}`;
        }
        return transacaoStr;
    }
    
}